### ISSUE TITLE
Using npm3

### ISSUE THREAD

**Issue by gilligan** [adnelson/nixfromnpm#62]
I see that `npm3` defaults to `false` in the top level `default.nix`. How can I actually use npm3 ? At first I thought I just need to set it to `true` but that is not the case I guess - i have to provide a tgz ? I am a bit confused about this ;-)

**Comment by adnelson** [comment]
So I used to have it use `npm3` by default, but unfortunately some upstream dependency of npm3 is currently failing to fetch because of problems with release-tag ranges in `semver-range` (by far this is the number 1 thing I need to fix with this project right now...). To get a working copy of `npm3`, first run

```bash
nixfromnpm -o /some/path -p npm
```

(or `-p npm%3` if you want to be explicit)

This will (probably, unless the situation is fixed by now) fail, at which point you can follow the steps in the README to fix the problem (i.e. manually npming the failed dependency, seeing what version gets fetched, fetching that with `nixfromnpm`, and updating the broken dependency). Once this is done, you can change the default of `npm3` to `true` in the top-level `default.nix` file, and everything should be ok.

I've already done this in the `nix-node-packages` repo, but if you want to have it work with a fresh package set, you'll have to do it yourself, until I fix the `semver-range` issue.

**Comment by gilligan** [comment]
Has anything happened on this front since?

**Comment by adnelson** [comment]
@gilligan 

Sorry I haven't really taken a look at this. I just generated npm3 and the error in the `requests` dependency (which requires a version of `form-data` that `semver-range` doesn't know how to satisfy) remains. Needless to say, `semver-range` has not yet been fixed, because I haven't spent time on it and still don't have a clear enough grasp of how semver handles version ranges that have release tags. 

So you can fix the broken upstream dependencies as described in the README. However, there's also a problem with some circular dependencies (es6-iterator, es6-symbol, d, and es5-ext all depend on each other). There are also instructions on how to fix this particular issue in the README. So long story short:

1. generate an npm3 expression by running the command I put in my above comment, then fix the broken `form-data` dependency in the `requests` library.
2. fix the 4 circularly-dependent packages I listed above.

This is a bit of a pain but you only need to do it once. After that point you will be able to use npm3 to build your node packages.

There are two other options which are worth considering:

1. Use `nix-node-packages` for your node packages. This already has a working version of npm3 including the above fixes, as well as many other packages, including those which had to be modified from what nixfromnpm generated. You might save yourself some work by using a fork of that repo.
2. Just don't use npm3, if you're not sure you need it.

Of course, at some point I hope to fix the semver-range issue, and the circular packages issue, but I think we both can tell that that's probably going to take longer than the above course of action. :wink:

**Comment by adnelson** [comment]
OK with the latest changes this should be able to be closed. I successfully generated a nix expression for the latest npm, and it built. 👻 👍

**Comment by gilligan** [comment]
Great! :+1:
